src/calling/haplotypes/virus.rs

```diff
@@ -206,3 +241,117 @@ impl Caller {
+    let mut new_event_posteriors = Vec::new();
+
+    // collect all lp haplotypes (with all distances)
+    let mut all_haplotypes: BTreeSet<Haplotype> = representatives.iter().cloned().collect();
+    for ((hap1, hap2), dist) in distance_matrix {
+        all_haplotypes.insert(hap1.clone());
+        all_haplotypes.insert(hap2.clone());
+    }
+    let all_haplotypes: Vec<Haplotype> = all_haplotypes.into_iter().collect();
+
+    dbg!(&all_haplotypes);
+
+    // create indices for every haplotype
+    let haplotype_indices: BTreeMap<Haplotype, usize> = all_haplotypes
+        .iter()
+        .enumerate()
+        .map(|(i, hap)| (hap.clone(), i))
+        .collect();
+    dbg!(&haplotype_indices);
+
+    for (fractions, logprob) in event_posteriors {
+        // create a new fractions vector by extending existing fractions initialized with 0.0
+        let mut expanded_fractions: Vec<AlleleFreq> =
+            vec![NotNan::new(0.0).unwrap(); all_haplotypes.len()];
+
+        // extend fractions using indices from representative haplotypes
+        for (i, &fraction) in fractions.iter().enumerate() {
+            if let Some(&idx) = haplotype_indices.get(&representatives[i]) {
+                expanded_fractions[idx] = fraction;
+            }
+        }
+
+        //push the row to event posteriors
+        new_event_posteriors.push((
+            HaplotypeFractions(expanded_fractions.clone()),
+            logprob.clone(),
+        ));
+
+        //push alternative solutions to event posteriors
+        for (haplotype, idx) in haplotype_indices.iter() {
+            // dbg!(&haplotype, &idx);
+            let fraction = expanded_fractions[*idx];
+            // dbg!(&fraction);
+            for ((hap1, hap2), &dist) in distance_matrix {
+                if dist == 0 && fraction > NotNan::new(0.0).unwrap() {
+                    // the second condition avoids having duplicate rows
+                    if hap1 == haplotype {
+                        if let (Some(idx1), Some(idx2)) =
+                            (haplotype_indices.get(hap1), haplotype_indices.get(hap2))
+                        {
+                            let mut alt1 = expanded_fractions.clone();
+                            alt1[*idx2] = fraction;
+                            alt1[*idx1] = NotNan::new(0.0).unwrap();
+                            new_event_posteriors.push((HaplotypeFractions(alt1), logprob.clone()));
+                        }
+                    } else if hap2 == haplotype {
+                        if let (Some(idx1), Some(idx2)) =
+                            (haplotype_indices.get(hap1), haplotype_indices.get(hap2))
+                        {
+                            let mut alt2 = expanded_fractions.clone();
+                            alt2[*idx1] = fraction;
+                            alt2[*idx2] = NotNan::new(0.0).unwrap();
+                            new_event_posteriors.push((HaplotypeFractions(alt2), logprob.clone()));
```

Comment:
The if-else looks redundant.

Reply:
I believe it's not redundant, as alternative rows will only be added if a haplotype in the distance matrix matches the current haplotype being processed in the loop.